### PR TITLE
Fix time entry report

### PR DESCRIPTION
**Pull request by shalapatil**
Notion: https://www.notion.so/saeloun/Time-entry-report-update-elastic-search-to-fetch-more-than-100-entries-5b4d36c71eb243d6b2c692d95a9d2384?pvs=4

What: Fixed the time entry report to get all the entries with pagination. 
When group by option is not passed, it fetches 10k records on single page. 
When group by option is passed, it fetches following number of groups and their respective data on single page
Clients -> 3
Projects -> 3
Users -> 5

Why: Time entry report was not fetching all the data due to limitation of page and the query pattern. 

Loom Video: https://www.loom.com/share/f4ac72f9323747409d50fc7adb511a9e

### CODE REVIEW

**Comment by vipulnsward** [comment]
@shalapatil can you check spec failures?

**Comment by supriya3105** [comment]
@shalapatil If the specs are now passing, please assign to Vipul for review. Let's try to get PR's approved faster.

File: app/services/reports/time_entries/page_service.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      @pagy_data, clients = pagy(clients_query, items: PER_PAGE[:clients], page:)
+      @es_filter = { client_id: clients.pluck(:id) }
+    end
+
+    def project_filter
+      @pagy_data, projects = pagy(projects_query, items: PER_PAGE[:projects], page:)
+      @es_filter = { project_id: projects.pluck(:id) }
+    end
+
+    def users_query
```
Comment (prasanthchaduvula) [inline app/services/reports/time_entries/page_service.rb:63]:
```suggestion
    is it possible to build something similar to Invoices::IndexService
```

File: app/services/reports/time_entries/report_service.rb [hunk 2]
```diff
@@ -1,54 +1,93 @@
...
 
-module Reports::TimeEntries
-  class ReportService
-    attr_reader :params, :is_download_request, :current_company
-
-    def initialize(params, current_company, download: false)
-      @params = params
-      @current_company = current_company
-      @is_download_request = is_download_request
+class Reports::TimeEntries::ReportService
```
Comment (prasanthchaduvula) [inline app/services/reports/time_entries/report_service.rb:3]: is it possible to build something similar to Invoices::IndexService

**Review by prasanthchaduvula — CHANGES_REQUESTED** [review]

**Comment by supriya3105** [comment]
@mayankagnihotri7 Please finish the review comments from Prasanth on this PR and get it closed.

File: app/services/duration_formatter.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+# Formats the duration passed in minutes to HH:MM format
+class DurationFormatter < ApplicationService
+  attr_reader :duration
+
+  def initialize(duration)
+    @duration = duration
+  end
+
+  def process
+    total_minutes = duration.to_i
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/services/duration_formatter.rb:12]: this is working code which is just refactored into service from model

File: app/services/reports/time_entries/report_service.rb (hunk 2 above)
Comment (shalapatil) [reply app/services/reports/time_entries/report_service.rb:3]: didn't get it, can you please explain?

File: app/models/client.rb [hunk 4]
```diff
@@ -41,6 +41,8 @@ class Client < ApplicationRecord
   after_discard :discard_projects
   after_commit :reindex_projects
 
+  scope :with_ids, -> (client_ids) { where(id: client_ids) if client_ids.present? }
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/models/client.rb]: IMO with_ids is clear name and conveying clearly that scope is to get for given ids.

File: app/models/project.rb [hunk 5]
```diff
@@ -40,6 +40,8 @@ class Project < ApplicationRecord
   after_discard :discard_project_members
   delegate :name, to: :client, prefix: true, allow_nil: true
 
+  scope :with_ids, -> (project_ids) { where(id: project_ids) if project_ids.present? }
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/models/project.rb]: Not changing it as per the above comment.

File: app/models/user.rb [hunk 6]
```diff
@@ -66,6 +66,7 @@ class User < ApplicationRecord
   rolify strict: true
 
   scope :with_kept_employments, -> { merge(Employment.kept) }
+  scope :with_ids, -> (user_ids) { where(id: user_ids) if user_ids.present? }
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/models/user.rb]: Not changing it as per the above comment

File: app/services/reports/time_entries/report_service.rb (hunk 2 above)
Comment (mayankagnihotri7) [reply app/services/reports/time_entries/report_service.rb:3]: This is handling the filter as well.

File: app/services/reports/time_entries/page_service.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      first: pagy_data.page == 1,
+      prev: pagy_data.prev,
+      next: pagy_data.next,
+      last: pagy_data.last
+    }
+  end
+
+  private
+
+    def es_filter_for_pagination
```
[the remark this answers is not in the document]
Comment (mayankagnihotri7) [reply app/services/reports/time_entries/page_service.rb:40]: Cannot currently do that due to rubocop warnings which come.

File: app/models/client.rb [hunk 8]
```diff
@@ -41,6 +41,8 @@ class Client < ApplicationRecord
   after_discard :discard_projects
   after_commit :reindex_projects
 
+  scope :with_specific_ids, -> (client_ids) { where(id: client_ids) if client_ids.present? }
```
Comment (prasanthchaduvula) [inline app/models/client.rb:44]: @mayankagnihotri7 Revert it to `:with_ids`

File: app/models/project.rb [hunk 9]
```diff
@@ -40,6 +40,8 @@ class Project < ApplicationRecord
   after_discard :discard_project_members
   delegate :name, to: :client, prefix: true, allow_nil: true
 
+  scope :with_specific_ids, -> (project_ids) { where(id: project_ids) if project_ids.present? }
```
Comment (prasanthchaduvula) [inline app/models/project.rb:43]: @mayankagnihotri7 Revert it to :with_ids

File: app/models/user.rb [hunk 10]
```diff
@@ -66,6 +66,7 @@ class User < ApplicationRecord
   rolify strict: true
 
   scope :with_kept_employments, -> { merge(Employment.kept) }
+  scope :with_specific_ids, -> (user_ids) { where(id: user_ids) if user_ids.present? }
```
Comment (prasanthchaduvula) [inline app/models/user.rb:69]: @mayankagnihotri7 Revert it to :with_ids

**Review by prasanthchaduvula — CHANGES_REQUESTED** [review]

**Review by prasanthchaduvula — APPROVED** [review]

**Comment by supriya3105** [comment]
@apoorv1316 Please test this functionality on local. If its working, please merge. Assigning it to you .

**Comment by apoorv1316** [comment]
@mayankagnihotri7 We are only displaying 3 client entries on a page but if we download the pdf I am also getting the entries for the client whose entries were not on the report page.  Please check loom: https://www.loom.com/share/e3c797352e9c4bf492e957d3c2ae926a

**Comment by mayankagnihotri7** [comment]
@apoorv1316 I've made the requested changes. Please review.
@prasanthchaduvula Can you please check the PR again as well? I've added frontend pagination.


https://user-images.githubusercontent.com/57438322/227517146-7564ee98-6354-415b-833d-84daef40e130.mp4

**Pull request merged**